### PR TITLE
Ssh conn info

### LINKED ISSUE #25: Basic SSH connection info collection
## Description:
As a security researcher I want to be able to collect SSH connection information so that I later can analyze the data.

**Difficulty:** 8

## Tasks:
- [x] Collect the connection data including connection data, login credentials tried and timestamp.
- [x] Design interface for logging module. (Benjamin & Sophia) https://github.com/Botnet-Honeypot/Honeypot/pull/29
- [x] Send to logging module (Log to console for now)

## Acceptance Criteria:
- [x] The server should be able to handle multiple connections
- [x] Is it possible to collect connection data, like login credentials tried and timestamp?
- [x] Is it possible for the collected information to be sent to the logging module (console for now)?

## Definition of Done:
- [x] All Acceptance Criteria of the user story are met
- [x] Code meets our Coding Standards.
- [x] Code is committed into GitHub.
- [x] Functional tests are performed. 
- [x] Code review is conducted by the majority of the group.
- [x] The code is covered by a minimum of 70% Unit Tests.
- [x] The code is documented (sphinx docstring).

### PR DESCRIPTION
**Pull request by ALX99**

### CODE REVIEW

File: frontend/frontend/__main__.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 
 import frontend.protocols.ssh as ssh
 
-key = paramiko.RSAKey(filename="./host.key")
-s = ssh.ConnectionManager(host_key=key, port=2222)
-s.listen()
+if __name__ == '__main__':
+    key = paramiko.RSAKey(filename="./host.key")
+    s = ssh.ConnectionManager(host_key=key, port=2222)
```
Comment (klondi) [inline frontend/frontend/__main__.py:8]: You may want to parametrize these later on, if you need to go down that rabbit hole, consider https://docs.python.org/3.9/library/argparse.html or one of the many libraries for that xD https://codeburst.io/building-beautiful-command-line-interfaces-with-python-26c7e1bb54df

File: frontend/frontend/honeylogger/__init__.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    """Begins to log a new SSH session.
+
+    :param src_address: The IP address of the instigating origin of the session.
+    :param src_port: The port number at the instigating origin.
+    :param dst_address: The public IP address of the honeypot.
+    :param dst_port: The port at the honeypot.
+    :return: An established SSH session.
+    """
+
+    session = ConsoleLogSSHSession(src_address=src_address,
```
Comment (klondi) [inline frontend/frontend/honeylogger/__init__.py]: Why do you need this proxy function? Wouldn't it work better to just call the constructor directly?

File: frontend/frontend/honeylogger/_console.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+handler = logging.StreamHandler(sys.stdout)
+formatter = logging.Formatter(
+    fmt='%(asctime)s %(levelname)-8s %(message)s', datefmt='%Y-%m-%d %H:%M:%S')
+handler.setFormatter(formatter)
+logger.addHandler(handler)
+
+
+class ConsoleLogSSHSession():
```
Comment (klondi) [inline frontend/frontend/honeylogger/_console.py:18]: If you plan to use this with a "with" statement you may want to define the __enter__ and __exit__ methods.

File: frontend/frontend/protocols/ssh.py [hunk 4]
```diff
@@ -57,30 +59,100 @@ def check_channel_shell_request(self, channel: paramiko.Channel) -> bool:
...
+        self._session.log_pty_request(term.decode("utf-8"), width, height, pixelwidth, pixelheight)
         return True
 
 
-# todo should be singleton if the class does what it says in the docstring
-class ConnectionManager():
-    """ConnectionManager contains logic for connecting incoming
-    SSH connections to instances of Server"""
-
+class ConnectionHandler(threading.Thread):
```
Comment (klondi) [inline frontend/frontend/protocols/ssh.py:66]: Not sure if Paramiko supports that, but you may want to check on coroutines to get that little extra edge :)

File: frontend/frontend/protocols/ssh.py [hunk 5]
```diff
@@ -124,40 +214,47 @@ def listen(self) -> None:
...
+            except Exception as exc:
+                # todo log
+                continue
 
-        # Negotiate a new SSH session
-        transport.add_server_key(self._host_key)
-        server = Server(self._usernames, self._passwords)
-        transport.start_server(server=server)
+            session = logger.begin_ssh_session(
+                src_address=ip_address(addr[0]),
```
Comment (klondi) [inline frontend/frontend/protocols/ssh.py:253]: I see a lot of argument passing like this. Wouldn't you benefit from containing all 4 inside its own class instead?

File: frontend/frontend/__main__.py (hunk 1 above)
Comment (ALX99) [reply frontend/frontend/__main__.py:8]: Good idea, we're thinking about creating a config file where we can set these types of settings

File: frontend/frontend/honeylogger/__init__.py (hunk 2 above)
Comment (ALX99) [reply frontend/frontend/honeylogger/__init__.py]: We have this so we can easily change between a ConsleLogSSHSession and PostgresLogSSHSession object depending if we want to log to console or the database

File: frontend/frontend/honeylogger/_console.py [hunk 6]
```diff
@@ -1,14 +1,22 @@
...
+
 from ._types import IPAddress
 
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+handler = logging.StreamHandler(sys.stdout)
+formatter = logging.Formatter(
+    fmt='%(asctime)s %(levelname)-8s %(message)s', datefmt='%Y-%m-%d %H:%M:%S')
+handler.setFormatter(formatter)
+logger.addHandler(handler)
```
Comment (Sannholm) [inline frontend/frontend/honeylogger/_console.py:15]: This configuration should probably be done in the root of the application or even in a config file. The module should not decide how logging is output. This is up to the person running the application.

**Review by osksvan — APPROVED** [review]
Looking good, all tests except the database pass, as it should be for now.

File: frontend/frontend/protocols/ssh.py [hunk 7]
```diff
@@ -57,30 +59,111 @@ def check_channel_shell_request(self, channel: paramiko.Channel) -> bool:
...
+
+        # todo wait for user to send shell request
+        # todo don't hardcode
+        chan.settimeout(2)
+        while True:
+            self._lock.acquire()
+            if not self._transport.active or self._terminate:
+                self._lock.release()
+                break
+            self._lock.release()
```
Comment (Sannholm) [inline frontend/frontend/protocols/ssh.py:135]: Might want to use a try-finally here to guaranteed release the lock. Looks a little dodgy to release inside if.

**Review by Sannholm — APPROVED** [review]
Looks good! Just some minor notes.

**Pull request merged**